src/couch/src/couch_users_db.erl

```diff
@@ -86,7 +90,63 @@ save_doc(#doc{body={Body}} = Doc) ->
         Doc#doc{body={Body4}};
     {_ClearPassword, Scheme} ->
         couch_log:error("[couch_httpd_auth] password_scheme value of '~p' is invalid.", [Scheme]),
-        throw({forbidden, "Server cannot hash passwords at this time."})
+        throw({forbidden, ?PASSWORD_SERVER_ERROR})
+    end.
+
+% Validate if a new password matches all RegExp in the password_reqexp setting.
+% Throws if not.
+validate_password(ClearPassword) ->
+    case config:get("couch_httpd_auth", "password_reqexp", "") of
+    "" ->
+        ok;
+    "[]" ->
+        ok;
+    ValidateConfig ->
+        case couch_util:parse_term(ValidateConfig) of
+        {ok, RegExpList} when is_list(RegExpList) -> 
+            % Check the password on every RegExp.
+            Loop = fun(RegExp) ->
+                check_password_with_regexp(ClearPassword, RegExp)
+            end,
+            lists:foreach(Loop, RegExpList),
+            ok;
+        {ok, NonListValue} ->
+            couch_log:error(
+                "[couch_httpd_auth] password_reqexp value of '~p' is invalid.",
+                [NonListValue]
+            ),
+            throw({forbidden, ?PASSWORD_SERVER_ERROR});
```

Comment:
Is `forbidden` the correct error here? I would argue 400 is more appropriate, since the password is "malformed" based on not passing the regexp

Reply:
If in [`save_doc/1`](https://github.com/apache/couchdb/blob/776f920021dd15c51355e9a904bdc5a2250b19d9/src/couch/src/couch_users_db.erl#L89) the `[couch_httpd_auth] password_scheme` config has an invalid value, the user will receive an 403 status. I also used 403 for all instances where a part of `[couch_httpd_auth] password_regexp` is wrong.
In the latest commit, I changed the status code for "malformed" password to 400.

Reply:
403 for “we can’t hash any PW because we are misconfigured” is fine because we used that before, although a 5xx code might be more appropriate, but no need to change all that now.

400 for “your password is bad” is ok too. Sadly, there is no 4xx for “you failed whatever custom validation I have going on here”, only ones specific to other HTTP behaviour like `Accept` or `Expect` headers,